File: tests/algorithms/transpiler/FS/conversions/ipv4_conversion.fs
// Generated 2025-08-07 10:31 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let hex_digits: string = "0123456789abcdef"
let rec split_by_dot (s: string) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable s = s
    try
        let mutable res: string array = [||]
        let mutable current: string = ""
        let mutable i: int = 0
        while i < (String.length (s)) do
            let c: string = string (s.[i])
            if c = "." then
                res <- Array.append res [|current|]
                current <- ""
            else
                current <- current + c
            i <- i + 1
        res <- Array.append res [|current|]
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let rec parse_decimal (s: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable s = s
    try
        if (String.length (s)) = 0 then
            failwith ("Invalid IPv4 address format")
        let mutable value: int = 0
        let mutable i: int = 0
        while i < (String.length (s)) do
            let c: string = string (s.[i])
            if (c < "0") || (c > "9") then
                failwith ("Invalid IPv4 address format")
            value <- (value * 10) + (int c)
            i <- i + 1
        __ret <- value
        raise Return
        __ret
    with
        | Return -> __ret
let rec to_hex2 (n: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable n = n
    try
        let mutable x: int = n
        let mutable res: string = ""
        while x > 0 do
            let d: int = ((x % 16 + 16) % 16)
            res <- (string (hex_digits.[d])) + res
            x <- x / 16
        while (String.length (res)) < 2 do
            res <- "0" + res
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let rec ipv4_to_decimal (ipv4_address: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable ipv4_address = ipv4_address
    try
        let mutable parts: string array = split_by_dot (ipv4_address)
        if (Seq.length (parts)) <> 4 then
            failwith ("Invalid IPv4 address format")
        let mutable result: int = 0
        let mutable i: int = 0
        while i < 4 do
            let oct: int = parse_decimal (_idx parts (i))
            if (oct < 0) || (oct > 255) then
                failwith ("Invalid IPv4 octet " + (_str (oct)))
            result <- (result * 256) + oct
            i <- i + 1
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
let rec alt_ipv4_to_decimal (ipv4_address: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable ipv4_address = ipv4_address
    try
        let mutable parts: string array = split_by_dot (ipv4_address)
        if (Seq.length (parts)) <> 4 then
            failwith ("Invalid IPv4 address format")
        let mutable hex_str: string = ""
        let mutable i: int = 0
        while i < 4 do
            let oct: int = parse_decimal (_idx parts (i))
            if (oct < 0) || (oct > 255) then
                failwith ("Invalid IPv4 octet " + (_str (oct)))
            hex_str <- hex_str + (to_hex2 (oct))
            i <- i + 1
        let mutable value: int = 0
        let mutable k: int = 0
        while k < (String.length (hex_str)) do
            let c: string = string (hex_str.[k])
            let mutable digit: int = 0 - 1
            let mutable j: int = 0
            while j < (String.length (hex_digits)) do
                if (string (hex_digits.[j])) = c then
                    digit <- j
                j <- j + 1
            if digit < 0 then
                failwith ("Invalid hex digit")
            value <- (value * 16) + digit
            k <- k + 1
        __ret <- value
        raise Return
        __ret
    with
        | Return -> __ret
let rec decimal_to_ipv4 (decimal_ipv4: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable decimal_ipv4 = decimal_ipv4
    try
        if (decimal_ipv4 < 0) || ((int64 decimal_ipv4) > 4294967295L) then
            failwith ("Invalid decimal IPv4 address")
        let mutable n: int = decimal_ipv4
        let mutable parts: string array = [||]
        let mutable i: int = 0
        while i < 4 do
            let octet: int = ((n % 256 + 256) % 256)
            parts <- Array.append parts [|_str (octet)|]
            n <- n / 256
            i <- i + 1
        let mutable res: string = ""
        let mutable j: int = (Seq.length (parts)) - 1
        while j >= 0 do
            res <- res + (_idx parts (j))
            if j > 0 then
                res <- res + "."
            j <- j - 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
printfn "%d" (ipv4_to_decimal ("192.168.0.1"))
printfn "%d" (ipv4_to_decimal ("10.0.0.255"))
printfn "%d" (alt_ipv4_to_decimal ("192.168.0.1"))
printfn "%d" (alt_ipv4_to_decimal ("10.0.0.255"))
printfn "%s" (decimal_to_ipv4 (int 3232235521L))
printfn "%s" (decimal_to_ipv4 (167772415))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/PHP/conversions/ipv4_conversion.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function _intdiv($a, $b) {
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
$__start_mem = memory_get_usage();
$__start = _now();
  $hex_digits = '0123456789abcdef';
  function split_by_dot($s) {
  global $hex_digits;
  $res = [];
  $current = '';
  $i = 0;
  while ($i < strlen($s)) {
  $c = substr($s, $i, $i + 1 - $i);
  if ($c == '.') {
  $res = _append($res, $current);
  $current = '';
} else {
  $current = $current . $c;
}
  $i = $i + 1;
};
  $res = _append($res, $current);
  return $res;
};
  function parse_decimal($s) {
  global $hex_digits;
  if (strlen($s) == 0) {
  $panic('Invalid IPv4 address format');
}
  $value = 0;
  $i = 0;
  while ($i < strlen($s)) {
  $c = substr($s, $i, $i + 1 - $i);
  if ($c < '0' || $c > '9') {
  $panic('Invalid IPv4 address format');
}
  $value = $value * 10 + (intval($c));
  $i = $i + 1;
};
  return $value;
};
  function to_hex2($n) {
  global $hex_digits;
  $x = $n;
  $res = '';
  while ($x > 0) {
  $d = $x % 16;
  $res = substr($hex_digits, $d, $d + 1 - $d) . $res;
  $x = _intdiv($x, 16);
};
  while (strlen($res) < 2) {
  $res = '0' . $res;
};
  return $res;
};
  function ipv4_to_decimal($ipv4_address) {
  global $hex_digits;
  $parts = split_by_dot($ipv4_address);
  if (count($parts) != 4) {
  $panic('Invalid IPv4 address format');
}
  $result = 0;
  $i = 0;
  while ($i < 4) {
  $oct = parse_decimal($parts[$i]);
  if ($oct < 0 || $oct > 255) {
  $panic('Invalid IPv4 octet ' . _str($oct));
}
  $result = $result * 256 + $oct;
  $i = $i + 1;
};
  return $result;
};
  function alt_ipv4_to_decimal($ipv4_address) {
  global $hex_digits;
  $parts = split_by_dot($ipv4_address);
  if (count($parts) != 4) {
  $panic('Invalid IPv4 address format');
}
  $hex_str = '';
  $i = 0;
  while ($i < 4) {
  $oct = parse_decimal($parts[$i]);
  if ($oct < 0 || $oct > 255) {
  $panic('Invalid IPv4 octet ' . _str($oct));
}
  $hex_str = $hex_str . to_hex2($oct);
  $i = $i + 1;
};
  $value = 0;
  $k = 0;
  while ($k < strlen($hex_str)) {
  $c = substr($hex_str, $k, $k + 1 - $k);
  $digit = 0 - 1;
  $j = 0;
  while ($j < strlen($hex_digits)) {
  if (substr($hex_digits, $j, $j + 1 - $j) == $c) {
  $digit = $j;
}
  $j = $j + 1;
};
  if ($digit < 0) {
  $panic('Invalid hex digit');
}
  $value = $value * 16 + $digit;
  $k = $k + 1;
};
  return $value;
};
  function decimal_to_ipv4($decimal_ipv4) {
  global $hex_digits;
  if ($decimal_ipv4 < 0 || $decimal_ipv4 > 4294967295) {
  $panic('Invalid decimal IPv4 address');
}
  $n = $decimal_ipv4;
  $parts = [];
  $i = 0;
  while ($i < 4) {
  $octet = $n % 256;
  $parts = _append($parts, _str($octet));
  $n = _intdiv($n, 256);
  $i = $i + 1;
};
  $res = '';
  $j = count($parts) - 1;
  while ($j >= 0) {
  $res = $res . $parts[$j];
  if ($j > 0) {
  $res = $res . '.';
}
  $j = $j - 1;
};
  return $res;
};
  echo rtrim(json_encode(ipv4_to_decimal('192.168.0.1'), 1344)), PHP_EOL;
  echo rtrim(json_encode(ipv4_to_decimal('10.0.0.255'), 1344)), PHP_EOL;
  echo rtrim(json_encode(alt_ipv4_to_decimal('192.168.0.1'), 1344)), PHP_EOL;
  echo rtrim(json_encode(alt_ipv4_to_decimal('10.0.0.255'), 1344)), PHP_EOL;
  echo rtrim(decimal_to_ipv4(3232235521)), PHP_EOL;
  echo rtrim(decimal_to_ipv4(167772415)), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
